File: backend/internal/model/poopLocation.go
package model

import "time"

type LocationType string

const (
	Regular    LocationType = "regular"
	PortaPotty LocationType = "porta potty"
	Outhouse   LocationType = "outhouse"
	Other      LocationType = "other"
)

// we can use a single binary number to encode the seasons
const (
	summer = 1 << iota
	fall
	winter
	spring
)

type Season string

const (
	Summer Season = "summer"
	Fall   Season = "fall"
	Winter Season = "winter"
	Spring Season = "spring"
)

// map the number (in binary) to a list of seasons
func GetSeasons(mask int) []Season {
	var seasons []Season
	if mask&summer != 0 {
		seasons = append(seasons, Summer)
	}
	if mask&fall != 0 {
		seasons = append(seasons, Fall)
	}
	if mask&winter != 0 {
		seasons = append(seasons, Winter)
	}
	if mask&spring != 0 {
		seasons = append(seasons, Spring)
	}
	return seasons
}

func SeasonMask(seasons []Season) int {
	total := 0
	for _, s := range seasons {
		switch s {
		case Summer:
			total += summer
		case Fall:
			total += fall
		case Winter:
			total += winter
		case Spring:
			total += spring
		}
	}
	return total
}

type PoopLocation struct {
	UUID         string       `json:"uuid"`
	Name         string       `json:"name"`
	Latitude     float64      `json:"latitude"`
	Longitude    float64      `json:"longitude"`
	FirstCreated time.Time    `json:"first_created"`
	LastModified time.Time    `json:"last_modified"`
	LocationType LocationType `json:"location_type"`
	Seasonal     bool         `json:"seasonal"`
	Seasons      []Season     `json:"seasons"`
	Accessible   bool         `json:"accessible"`
	Upvotes      int          `json:"upvotes"`
	DownVotes    int          `json:"downvotes"`
}

File: frontend/lib/model/poop_location.dart
import 'package:latlong2/latlong.dart';
import 'package:flutter/material.dart';

class PoopLocation {
  String uuid;
  String name;
  double latitude;
  double longitude;
  String firstCreated;
  String lastModified;
  String locationType;
  bool seasonal;
  List<String> seasons;
  bool accessible;
  int upvotes;
  int downvotes;

  PoopLocation({
    required this.uuid,
    required this.name,
    required this.latitude,
    required this.longitude,
    required this.firstCreated,
    required this.lastModified,
    required this.locationType,
    required this.seasonal,
    required this.seasons,
    required this.accessible,
    required this.upvotes,
    required this.downvotes,
  });

  // return the lat/lng coord
  LatLng location() {
    return LatLng(latitude, longitude);
  }

  MaterialColor color() {
    switch (locationType) {
      case "regular":
        return Colors.red;
      case "porta potty":
        return Colors.green;
      case "outhouse":
        return Colors.brown;
      case "other":
      default:
        return Colors.blue;
    }
  }

  List<Widget> seasonsIcons() {
    Map<String, Widget> seasonsMap = {
      "summer": const Column(children: [Icon(Icons.sunny), Text("summer")]),
      "spring": const Column(children: [Icon(Icons.eco), Text("spring")]),
      "winter": const Column(children: [Icon(Icons.ac_unit), Text("winter")]),
      "fall": const Column(children: [Icon(Icons.nature), Text("fall")]),
    };
    return seasons
        .map((season) => seasonsMap[season] ?? const Icon(Icons.help_outline))
        .toList();
  }

  // A method to convert the PoopLocation instance to JSON
  Map<String, dynamic> toJson() {
    return {
      'uuid': uuid,
      'name': name,
      'latitude': latitude,
      'longitude': longitude,
      //'first_created': firstCreated,
      //'last_modified': lastModified,
      'location_type': locationType,
      'seasonal': seasonal,
      'seasons': seasons,
      'accessible': accessible,
      'upvotes': upvotes,
      'downvotes': downvotes,
    };
  }

  factory PoopLocation.fromJson(Map<String, dynamic> json) {
    return PoopLocation(
      uuid: json['uuid'],
      name: json['name'],
      latitude: json['latitude'],
      longitude: json['longitude'],
      firstCreated: json['first_created'],
      lastModified: json['last_modified'],
      locationType: json['location_type'],
      seasonal: json['seasonal'],
      seasons: json['seasons'],
      accessible: json['accessible'],
      upvotes: json['upvotes'],
      downvotes: json['downvotes'],
    );
  }
}
